### ISSUE TITLE
Update to modern QEMU!

### ISSUE THREAD

**Issue by AndrewFasano** [panda-re/panda#570]
We've been talking about this for a bit but haven't started work on it. Creating this issue to track progress.

We're currently forked off of Qemu at version 2.9.1. We should update to 4+. At the time of writing, the latest version if 4.1. 

We'll likely need to disable MTTCG to avoid significantly changing the record/replay model.

The main tasks I see for now are:
- [ ]  Actually do the **git merge** and handle merge conflicts to create a (likely broken) branch with all the commits
- [ ] Get **callbacks** to run
- [ ] Capture correct **recordings**
- [ ] **Replay** recordings correctly
- [ ] **Library mode**: build qemu as a library 
- [ ] **Pypanda** testing
- [ ] Extensive **testing**

**Comment by janbbeck** [comment]
Just to give it my 2 cents:
I recently upgraded qira from qemu 2.x to qemu 4.0  instead of 4.1. At least for that effort, the changes after 4.0 were getting so severe that porting to 4.0 before going to 4.1 made sense.
This may, of course, not be true for PANDA, but it may be easier to port PANDA to some intermediate version of qemu than going to 4.1 all at once.

**Comment by hanetzer** [comment]
judging by the amount of monkeypatching I needed to do to get pandas as it currently
exists to build on my system, I almost think it would be easier to collect what is different
between pandas and the qemu version its based on and just rewrite it against qemu4.

**Comment by AndrewFasano** [comment]
@janbbeck Thanks for the tip. 4.0 sounds like it might be a more realistic target.

@hanetzer It's unlikely that any of us have time to do a full rewrite given that there are 2k commits to PANDA since we forked. If we had time, that would certainly be the cleanest way. Also- you shouldn't need to do any monkeypatching- if you've had build issues can you open an issue? Our CI has been able to build PANDA on clean machines without any problems.

**Comment by hanetzer** [comment]
@AndrewFasano I'm running on gentoo ~amd64 (read: bleeding on the edge),
so there's a lot of new gcc diagnostics and such that make the build fail in creative ways.

While I have a human looking at this, I'll open an issue about what I'm trying to
do and what is/isn't/idk working.

**Comment by AndrewFasano** [comment]
@hanetzer Ah, okay, that makes more sense then. We just fixed some gcc7-related errors in the last few weeks but that's still probably not new enough for you.

**Comment by hanetzer** [comment]
@AndrewFasano yeah. I'm currently working under a deadline and this
is the last tool I can think of to help me do what I want, but once I either
fail the task or succeed I'm more than willing to help put in the work on
the update. Think you could eyeball #577 and toss me some suggestions?

**Comment by janbbeck** [comment]
fwiw, panda builds just fine for me with gcc 7 8 ad 9 under ubuntu 19.04, as long as werror is suppressed. no monkeying necessary, using this type of configure:
```
./configure --target-list=x86_64-softmmu,i386-softmmu,arm-softmmu,ppc-softmmu --prefix=/home/jan/Downloads/panda/panda/scripts/panda/build/install --python=/usr/bin/python2 --disable-vhost-net --extra-cflags=-DXC_WANT_COMPAT_DEVICEMODEL_API --extra-cflags=-DOSI_PROC_EVENTS --extra-cflags=-DOSI_MAX_PROC=256 --extra-cflags=-DOSI_LINUX_PSDEBUG --extra-cflags=-Wformat-truncation=0 --disable-werror --cc=gcc-9 --cxx=g++-9 --host-cc=gcc-9
```
- just change to compiler of your choice.

I just rebuilt panda with gcc 7, 8 and 9 and booted each time into a ubuntu live dvd and noticed no issues. HTH

**Comment by nathanjackson** [comment]
Bumping versions is hard, but not impossible. About a year ago, I worked on bringing PANDA up to 2.9.1 from an in-development version of QEMU 2.9. That itself was an undertaking, but mostly because I had to figure out what MTTCG changes were made and favor the PANDA version.

I would recommend bumping versions incrementally. I know you really want QEMU 4 (or maybe even now, QEMU 5) but I suspect this won't be easy.

I'd follow this approach, as it's what I did to bring PANDA to 2.9.1.

1. Pick a version, recommend 2.10.2.
2. Make a list of commits that need to be merged in.
3. Merge and bisect over the commits to be merged to see where stuff breaks.
4. Test, etc.
5. Rinse and repeat for 2.11.2, ... 5.0, etc.

While this is somewhat tedious there's probably a fair amount of automation you could do and you'll be in a working state at each step of the way and avoid huge merge conflicts.

**Comment by janbbeck** [comment]
I would like to point out that one of the major reasons I stopped at 4.0.0 for the qira port is that after that they added a plugin architecture of some sort that greatly changed the source files. It sure appered tailored for the sort of thing qira and panda do. It looked a lot to me like the correct thing to do from that point forward is re-do qira as a plugin - and that was more than I was willing to bite off :p
I am bringing this up, because maybe - just maybe - the qemu plugin stuff is actually the way panda should be done in newer versions of qemu. If that is the case, I am sure there will be no end of discussion between porting piecemeal or just biting the bullet and making a plugin.

But I think it's worth a look before deciding to go piece by piece to 4.0.0 and beyond.

Thoughts?

**Comment by janbbeck** [comment]
here more info:
https://github.com/qemu/qemu/blob/master/docs/devel/tcg-plugins.rst

**Comment by AndrewFasano** [comment]
I've looked at the tcg plugins a little and I'm really excited that qemu is finally starting to support using it for analysis, but I think there's still a long way to go before it could support everything we use/provide with panda. The biggest issue I see for now is that the TCG plugins only support passive observation so you can't modify a system during its execution.

**Comment by janbbeck** [comment]
That is good to know! I suspect the code changes in qemu to support the plugins make porting panda painful....

**Comment by AndrewFasano** [comment]
I guess most of our plugins do passive analysis, but a lot of my research involves mutating guest state so I'm probably a bit biased when I consider that shortcoming to be a dealbreaker.

**Comment by janbbeck** [comment]
Is there some particular shortcoming that necessitates to move away from the current qemu base?

**Comment by AndrewFasano** [comment]
@mariusmue first brought it up so he might want to chime in. I mainly want support for more machine types and whatever stability/performance fixes they've made.

**Comment by mariusmue** [comment]
Hi all,
In my opinion, an update to a newer version of Qemu benefits foremost emulation and analysis of non x86-based systems. As @AndrewFasano mentioned, there are just more machines, as well as architectures implemented. Furthermore, when it comes to avatar-related changes, being on modern versions of QEMU would it make easier to sync changes between the two frameworks, but this is just a minor point.

MTTCG is luckily guarded well behind preprocessor macros, so it should be easy to keep it deactivated. Furthermore, I think somewhen around qemu 3.0 was a tremendous change/cleanup in the QAPI, I personally like the new organization better.

When it comes to tcg-plugins, yes, they are great for *passive* monitoring of VMState. In case PANDA wants to enforce a strict separation between record and analysis, I would suggest that recording is re-implemented on top of this API (if possible), as this would allow recording on *stock-builds* of qemu as drop-in solution. In theory, this should even allow for records independent of the QEMU version, but I think there may various problems arise in praxis. (E.g., different peripheral implementations.)
In any case, the actual analysis/replay instances of PANDA would still need to hook at various in the codebase, but distinguishing whether we are in record or replay mode should be an artifact of the past, allowing for a cleaner codebase, easier to migrate to upcoming versions of qemu as well.
Hence, if this changes are going to happen, I would not plead for a complete rewrite of PANDA, but for identifying and minimizing the locations PANDA actually hooks into QEMU's core logic.

**Comment by glueckself** [comment]
I'm merging Qemu 5.0 rc1 into PANDA as part of my thesis.
I hope that going from rc1 to the release version wont be a big problem.

I jumped right into it (I do question myself if that was the smart way) and I have the conflicts down to cpus.c and softmmu_template.h, and whatever errors the compiler will throw at me when I first build it. There were a few parts that were drastically changed, mostly around the TCG.

I'm waiting on the reply from my supervisor about how and when we can publish the source, but I guess I'll know more next week and by then I think I'll also have the merge commit.

**Comment by AndrewFasano** [comment]
Hey @glueckself, that's great to hear! After you finish the merge, I'd expect there to be lots of bugs (unless you're really good at merging) as PANDA usually needs additional updates when core QEMU things change. Once you have a merge commit, I think there are a number of us who would be willing to help track down bugs if you're able to share!

**Comment by m000** [comment]
Some thoughts about the (welcome) migration to QEMU 4 codebase.

Since PANDA will continue following the development of QEMU, maybe it would be helpful to label PANDA releases based on the underlying QEMU codebase? E.g. current version would be PANDA2 (based on QEMU 2.x) codebase. Next version would be PANDA4 (based on QEMU 4.x). This would make it a bit easier to discuss issues while both versions are in use.

I can see this issue growing longer and longer. Maybe a separate branch should be created while stabilizing/working out bugs with the QEMU4-based panda? Then issues related to the branch can be reported individually. A new QEMU4 or PANDA4 label for the issue tracker can be used to filter issues quickly. 

Finally, it would probably be good to also announce a draft time-plan for the deprecation of the current code-base. This would encourage the community to migrate any incompatible code to the new version.

**Comment by glueckself** [comment]
If my supervisor is ok with me publishing the code on Github, I can fork this repository. The fork would then provide a separate issue tracker. When the migration is completed, you could merge my fork back here.
Please note that I'm migrating PANDA to Qemu 5.

**Comment by AndrewFasano** [comment]
@m000 I like your suggestions - then we get to jump from PANDA2 to PANDA4 and skip all the work of PANDA3 ;) I think we should wait until we have the new version at least partly working before we plan any deprecation timelines.

I think we'll take a look at @glueckself's code if/when that's available and then pull it into a branch on this repo, instead of tracking it in a separate repo. QEMU 5 sounds great if you're able to get it to work. Then we can go right up to PANDA5 :)

**Comment by glueckself** [comment]
I've now uploaded the [branch containing the merge](https://github.com/glueckself/panda/tree/panda-50), however, I'm still fixing compiler errors. I have never merged anything this big before, so I expect there will be some mistakes in there. I would appreciate any feedback. :)
I also haven't looked into testing of Qemu and PANDA yet, I think that'll be the next step after getting it to compile.

Regarding the checklist: my goal is to get record/replay running (and probably only for i386 and arm). After that, I'll probably have to switch over to my thesis (btw, it's only a BSc thesis).

**Comment by nathanjackson** [comment]
There will be a public regression testing framework available soon.

I just looked at your branch: 23610 commits ahead, 654 commits behind panda-re:master.... Godspeed.

**Comment by glueckself** [comment]
I've got qemu-system-i386 to build, however, some of the "fixes" I made are... a bit ugly (especially ef6a13e1f81cf029fc2e953c2406622717157b45). It manages to boot [PC-MOS/386](https://www.qemu-advent-calendar.org/2018/download/day05.tar.xz). And, with a workaround, Linux. But only a [test image](https://github.com/palmercluff/qemu-images/tree/master/test-images/linux-test). Debian hangs on some udev soft lockups.

Issues I've discovered so far:
* It's very slow. I've tried to look into it, but I couldn't find anything.
* The main thread has 100% CPU load (might be what causes it to be slow).
* The cdrom Device timeouts. This causes Linux to fail to boot. Workaround is to start qemu with `-nodefaults -vga std` (I'm not sure if there is a way to remove only the cdrom drive). Haven't looked into it yet.
* Soft lock ups in the guest.
* It crashes when starting with `-llvm` or on `begin_record`.
* The qemu `tests/qtest/boot-serial-test` hangs.

I've marked some places where I don't understand what's going on (or what should be going on) with "//TODO: panda:". I'll try to sort out as many as I can. Also, there are a lot of warnings. I haven't looked into them yet, probably there are bad ones in there.

My next goal is to get everything to compile and clean up the warnings and TODOs.

I have problems with the C/C++ mixing. Qemu started to use some C specific stuff in (e.g. `__builtin_types_compatible_p()` in `include/qemu/atomic.h` ) and g++ doesn't support that. I'm not sure if C-linkages can solve that (i.e. if there is a `extern "C"` missing somewhere) or if that has to be implemented for C++. Would it be possible for someone to support me there?

**Comment by nathanjackson** [comment]
@glueckself I would highly recommend incrementally merging, even within a QEMU version. I would try getting PANDA to the next released version of QEMU first (2.10.X I think). Tracking down these issues will not be easy, certainly not something where someone on the internet can just point you in the right direction.

**Comment by glueckself** [comment]
@nathanjackson do you think that I'm on a dead end here? Because I've got it to build everything by now and my next step would be to clean stuff up. I guess most of the issues are somewhere near a "//TODO: panda:" comment and some are pointed to by compiler warnings that I've ignored for now. I don't think I would resolve most of those conflicts any better if they came one-by-one instead of all-at-one.
To be honest, I would hate to have to throw everything away (however, I do prefer to throw it away instead of looking for bugs for the next two years).

**Comment by nathanjackson** [comment]
I'm hesitant to say you're on a "dead end", but the issues you've listed would indicate you're running into complicated CPU loop\iothread issues. Where you start debugging is anyone's guess and it doesn't sound like you've even started looking at record/replay. Debugging this will be extremely difficult because QEMU is complicated and the code is multithreaded.

Since you're merging in a huge number of commits, how will you know which ones are causing problem(s) if you don't incrementally merge? How do you know your conflict resolution was correct for one issue while debugging another? Bisecting won't work well here because you've probably introduced more than one problem into the code. You need to do the opposite: introduce some commits -> test, then rinse and repeat. There are a large number of commits that should apply without you having to resolve conflicts, I promise this isn't as bad as it sounds.

If you try the incremental route, I don't think you have to throw everything away per se. By attempting the big merge, you've already learned some things about the code. You'll be able to use that knowledge as you're incrementally merging.

I don't want to discourage you from doing this, I think everyone including me wants PANDA to be on a newer QEMU. I'm just saying how I would go about doing this. If you think that you can make it work with one big merge, by all means do it.

**Comment by glueckself** [comment]
Do you think it makes sense for me to try and resolve the places I've marked and maybe dig deeper into one or two of issues? Maybe over the next two weeks or so, and if after that it's still a huge mess, then I could try the incremental route?

But you're right, bisecting becomes an issue then. Would such a merge be acceptable for maintaining anyway?

No worries, I'm thankful for the feedback :) I've never done anything like this before, so I expect a set-back or two :)

**Comment by nathanjackson** [comment]
> Do you think it makes sense for me to try and resolve the places I've marked and maybe dig deeper into one or two of issues? Maybe over the next two weeks or so, and if after that it's still a huge mess, then I could try the incremental route?

Honestly, no. I think any **one** of those issues you mentioned is probably more than two weeks of work if you were doing this full-time. You mentioned a bachelor's thesis. My guess is that you have other work to do as well. I think an incremental merge will get you almost immediate results.
 
> 
> But you're right, bisecting becomes an issue then. Would such a merge be acceptable for maintaining anyway?

If we do this incrementally or in one commit, I think bisecting will be hard for awhile anyway because of the changes being made. I wouldn't let that stop you in either case. I was really saying during the merge, figuring out which upstream QEMU change broke something would be next to impossible with one giant merge.

**Comment by janbbeck** [comment]
@glueckself FWIW, I could not agree more with Nathan about everything he has told you.

**Comment by AndrewFasano** [comment]
I spent more time than I'd like to admit over the past few days trying to merge qemu 2.10 into PANDA. The merge itself was fairly straightforward and it just took a few hours to handle the conflicts and get it to build.

But, unsurprisingly, the built version doesn't actually work. I've spent ~2 days debugging it and I'm still stuck on an error about "Trying to execute code outside RAM or ROM at 0x38000". When I take a trace of instructions executed with `-d in_asm`, I see about 6k instructions run before the error and they're the same as what qemu 2.10 execute. I think the error must be related to how memory regions are initialized and how the kernel is loaded into memory, but I'm not sure.

I was testing x86_64 with a Linux 4.9.99 kernel I built from source:
```
$ ./x86_64-softmmu/panda-system-x86_64  -kernel ./linux-4.9.99/arch/x86_64/boot/bzImage -nographic -append "console=ttyS0"

panda-system-x86_64: Trying to execute code outside RAM or ROM at 0x0000000000038000
This usually means one of the following happened:

(1) You told QEMU to execute a kernel for the wrong machine type, and it crashed on startup (eg trying to run a raspberry pi kernel on a versatilepb QEMU machine)
(2) You didn't give QEMU a kernel or BIOS filename at all, and QEMU executed a ROM full of no-op instructions until it fell off the end
(3) Your guest kernel has a bug and crashed by jumping off into nowhere

This is almost always one of the first two, so check your command line and that you are using the right type of kernel for this machine.
If you think option (3) is likely then you can try debugging your guest with the -d debug options; in particular -d guest_errors will cause the log to include a dump of the guest register state at this point.

Execution cannot continue; stopping here.
```

I tried redoing the merge from scratch but ran into the same issue. Code from the second merge (likely the better of the two) is on the [panda_next2](https://github.com/panda-re/panda/tree/panda_next2) branch.

Anyone interested in helping me debug?

**Comment by glueckself** [comment]
@AndrewFasano The initial SMI handler is located at 0x38000. I also have a problem with the SMM, but I'm not sure if it's exactly the same as yours. Also, I'm working with i386.
You can "big-picture" test it with `-machine smm=off`. If doesn't boot there we definitely have different problems.

In my case, it's that the SMM address space (asidx=1) doesn't see the changes done by the "normal" (not sure what it's called) address space (asidx=0). However, Seabios writes a few trampoline instructions at 0x38000 and after some time it triggers the SMI. The xp monitor command, "outer"-gdb (i.e. gdb --args -- ./qemu-system-i386 ...) + gpa2hva monitor command, "inner"-gdb (i.e. target remote :1234), the in_asm log all see `0x8c 0xc8 ...` (the instructions written by Seabios), but the TCG gets 0x00 when it wants to translate while the CPU is in SMM.
You can verify that by:
* starting qemu with outer-gdb and `-S`
* setting inner-gdb breakpoint to 0x38000 (connected to the debugged qemu process)
* letting inner-gdb continue (should break at 0x38000)
* breaking outer-gdb, and setting a breakpoint on load_helper . continue.
* continue inner-gdb.
* now the load_helper breakpoint shoult hit with addr being 0x38000. in my tests it also got hit by tcg-runtime code (check the callstack), if that's the case continue it. you need it from the translator.
* if its from the translator, `finish` it. if it returns 0, its bad. if it returns 0x8c, it's, well also bad, but then we have different bugs.

I was sick last week, but I've managed to talk to my supervisor. I'll continue with the big merge for 1-2 weeks and see where it leads me.
I've got the cdrom bug (missed the change from number-of-bytes to MO_ and it caused an endianess error for some in/out instructions) so that's solved.

Also the performance got better (possibly fixed in one a warnings cleanup). I manage to boot a Win7 and a Debian Buster without any hick-ups (beside the smm=off). I haven't pushed the latest version to Github yet because it's in a "temporary & dirty" branch, let me know if you want it.
`begin_record` does something, but the replay doesn't work because instruction mismatch. However, I haven't looked into it yet if records generates wrong output or if replay doesn't work.

**Comment by glueckself** [comment]
@AndrewFasano I see that you have a commit to disable SMM. Is that a workaround for now to be able to test everything else or do you plan to keep it that way? Please let me know, because if it's the later, then I'll also just disable it and focus on the record/replay issue. Otherwise I'll keep digging.

**Comment by AndrewFasano** [comment]
Hey, thanks for the info, that's super helpful! I just started looking at it last night and found that I was getting a little farther with SMM disabled, but then I get a segfault due to some memory corruption. I'll take a closer look at the SMM stuff and report back whatever I figure out.

**Comment by glueckself** [comment]
Ok, then I'll also keep looking. My next step will be to see what `address_space_translate_for_iotlb()` returns in vanilla Qemu while in SMM and see from there.
I'll also report my findings.

On a different note: have you tried arm-softmmu? Mine has a lockup (I think that what its called when the HardFaultHandler keeps hardfaulting).

**Comment by glueckself** [comment]
Vanilla Qemu has a tlb_hit in `#4 [...] get_page_addr_code_hostp [...]` of the "bad" callstack above, and that's where it takes on a different path.

The write to 0x38000 in bios-256k.bin happens in the guest at 0x000ec8a8, btw.

I currently have two approaches:

In the Seabios source, the [SMM setup](https://github.com/coreboot/seabios/blob/master/src/fw/smm.c#L180) enables some [memory window](https://github.com/coreboot/seabios/blob/master/src/fw/smm.c#L188). I think one approach is to see if that fails, it may explain why the two address spaces see something different.

The other approach is to find out why the TLB entry received in `get_page_addr_code_hostp()` is different:
* panda-5.0:
```
gdb$ p *tlb_entry(env, mmu_idx, addr)
{{{addr_read = `*0x38200, addr_write = 0x38200, addr_code = 0x38200, addend = 0xfffffffffffc8000}, 
    dummy = "\000\202\003\000\000\202\003\000\000\202\003\000\000\000\000\000\000\200\374\377\377\377\377\377\340k&\360\377\017\000"}}
```
* qemu-5.0:
```
gdb$ p *tlb_entry(env, mmu_idx, addr)
{{{addr_read = 0x38000, addr_write = 0x38400, addr_code = 0x38000, addend = 0x7fff89600000}, dummy = "\000\200\003\000\000\204\003\000\000\200\003\000\360\377\017\000\000\000`\211\377\177\000\000\b<\367UUU\000"}}
```
(mmu_idx = 2 and addr = 0x38000 for both cases).

**Comment by glueckself** [comment]
SMM is fixed.
It's caused by the `rr_listener` that points to the `rr_mem_region_{added,deleted}_cb` callbacks in `cpu_address_space_init()`. The same `rr_listener` struct is registered multiple times which messes something up. I've workarounded/fixed this by `calloc()`ing it on the heap for each call:
```diff
diff --git a/exec.c b/exec.c
index 0837fdd840..927e5b9480 100644
--- a/exec.c
+++ b/exec.c
@@ -851,8 +851,6 @@ CPUState *qemu_get_cpu(int index)
 
 #if !defined(CONFIG_USER_ONLY)
 
-MemoryListener rr_listener;
-
 static RR_mem_type rr_mem_region_type(MemoryRegion* mr) {
     RR_mem_type mtype = RR_MEM_UNKNOWN;
     if (!(memory_region_is_ram(mr) || memory_region_is_romd(mr))) {
@@ -913,10 +911,12 @@ void cpu_address_space_init(CPUState *cpu, int asidx,
         newas->tcg_as_listener.commit = tcg_commit;
         memory_listener_register(&newas->tcg_as_listener, as);
 
+        //TODO: panda: maybe use something other than calloc?
+        MemoryListener *rr_listener = calloc(sizeof(MemoryListener), 1);
         // PANDA Record and Replay
-        rr_listener.region_add = rr_mem_region_added_cb;
-        rr_listener.region_del = rr_mem_region_deleted_cb;
-        memory_listener_register(&rr_listener, as);
+        rr_listener->region_add = rr_mem_region_added_cb;
+        rr_listener->region_del = rr_mem_region_deleted_cb;
+        memory_listener_register(rr_listener, as);
     }
 }
```
I'll push it soon.

So far, I would say that qemu basics run. But I haven't tried network yet.

However, running it with `-llvm` caused some bugs because the old S2E code in `panda/llvm/tcg-llvm.cpp` doesn't match the newer Qemu TCG. I've fixed or workarounded some of those, but now I'm back to SMM (see below) and I don't understand the message:
```
LLVM ERROR: Cannot select: 0x7fda00dc8520: i64 = X86ISD::WrapperRIP 0x7fda0192dad0 [ID=8]
  0x7fda0192dad0: i64 = TargetGlobalTLSAddress<%struct.rcu_reader_data* @rcu_reader> 0 [TF=10] [ID=6]
In function: log_cpu_state_mask
```
Here is the full llvm_ir log: https://gist.github.com/glueckself/b61ccdc07a3c2456cdcc11401db9908c

It happens on the `rsm` instruction (the instruction to leave SMM).
If I disable SMM (`-machine smm=off`) another error occurs (`ERROR: unknown TCG micro operation 'extract2_i32'`) but I think I can handle that one.

**Comment by tibersam** [comment]
Hi all,

is the update to Qemu 4.x/5.x currently undergoing or did it went stale?
What are the future plans?

I might be able to help but i am unsure about the current state?
Is the fork of glueckself the current dev branch?

**Comment by AndrewFasano** [comment]
Hi @tibersam,

@glueckself had been trying to directly merge PANDA with QEMU version 5.x and it seems like that might've been too big of a challenge since we haven't heard anything about it in a few months.

@nathanjackson is the only person in this thread who has successfully updated PANDA to a newer QEMU version, so I figure his advice of merging/fixing/testing a single minor version update at a time is probably the best plan.

With that in mind, I had tried merging in upstream version 2.10 on the `panda_next2` branch. I went though the conflicts with reasonable care and applied the SMM patch @glueckself posted above, but I was still running into a segfault that I wasn't able to track down.

I think our future plans are to incrementally merge in minor versions one at a time but it hasn't been a high priority for our core development team. If you're willing to work on it, we'd be happy to help you with it however we can.

**Comment by nathanjackson** [comment]
Just a quick question for the people that want a newer QEMU. Are there specific features from QEMU 4+ that are needed? Has anyone considered that back porting new architectures/boards to PANDA might actually be easier than upgrading PANDA itself?

Upgrading the QEMU baseline should be possible, but it will involve dropping commits or using the old version anyway, particularly around the CPU loop.

**Comment by janbbeck** [comment]
> Just a quick question for the people that want a newer QEMU. Are there specific features from QEMU 4+ that are needed? Has anyone considered that back porting new architectures/boards to PANDA might actually be easier than upgrading PANDA itself?
> 
> Upgrading the QEMU baseline should be possible, but it will involve dropping commits or using the old version anyway, particularly around the CPU loop.

I think that is a very good question to ask before embarking on a process such as this.....

Can anyone think of some nice-to have feature that qemu introduced in later versions? That would probably also point to a different version to strive for. For example if useful feature A was introduced in qemu x.y.z, then that may be a more reasonable target than qemu 4+

**Comment by AndrewFasano** [comment]
Still working on this! I've been exploring re-implementing some of the core PANDA features on top of QEMU 6 in https://github.com/andrewfasano/futurepanda by expanding their plugin interface to add support for various PANDA capabilities. So far I've implemented PPP-style callbacks and access to reading guest registers. You can see example plugins in the panda directory, e.g., [syscalls3](https://github.com/AndrewFasano/futurepanda/blob/main/panda/syscalls3.c).

Not sure if that will pan out, and a lot of key PANDA features (record/replay, LLVM, taint, pypanda/panda-rs) are out of scope, at least for now. But maybe there'd be be some value in supporting two versions of PANDA concurrently: 1) this version where you get more features, but it's based off older qemu, and 2) a version designed for live-analyses which is up to date with upstream, but lacks some features. Moving plugins out of tree (#1088) might mean that we could support the same plugins across both versions (though that would be a fair amount of work, given how upstream has reimplemented/renamed a bunch of APIs and we'd need to build some shim layers).

**Comment by rjzak** [comment]
This might resolve #1077, extern C vs. glib templates. Broken in Pop_OS (Ubuntu) 21.04 and Ubuntu 21.10.

**Comment by rjzak** [comment]
Bump.... still an issue even if stale.

**Comment by Manouchehri** [comment]
I know this might not be too helpful anymore, but if you're trying to merge in a ton of commits, I would highly recommend using git bisect with a test script (e.g. do a clean full build and run a test target) to see which changes break functionality in PANDA.

**Comment by fishfacegit** [comment]
Hi,
i just wanted to ask what the status of the qemu upgrade is?
Since malware have greatly improved on detecting VMs, it would be great to have a modern qemu to counter rdtsc reads on vm exit and similar instructions via Hyper-V Enlightenment.
From panda-re perspective, as far as i understand, you cannot record what is not executed.
Best Regards

**Comment by AndrewFasano** [comment]
We are actively developing a new version of PANDA atop qemu 8 using TCG plugins. We’re working with upstream to merge a new version of the PPP interface to allow plugins to interact with one another which we see as the first major blocker for PANDA-like functionality. After we hopefully get that merged, we’ll need to upstream patches to allow plugins to access guest registers and memory. Then we can start porting and upstreaming plugins as well. We currently have syscalls, osi_linux, and stringsearch ported to the new interface. And a limited pypanda interface. We haven’t yet tackled the record/replay system or LLVM IR as required by the taint system.

If you’re interested in joining the discussion with the QEMU developers, our latest patch series is at the following link. So far the discussion has only been between us and the maintainer, so other perspectives might be appreciated.

https://www.mail-archive.com/qemu-devel@nongnu.org/msg926042.html

Unfortunately our funding for this work runs out in a few days so we’re not sure what the future of this will be, but the work we've done so far is available [here](https://github.com/panda-re/qemu/tree/ppp_beta_with_plugins), and if upstream merges the QPP stuff we’ll try getting some of our API changes and plugins upstreamed too

**Comment by ghost** [comment]
@AndrewFasano I want to make sure that the panda is based on the branch of the 2.9.1 stable version of the qemu warehouse?
After looking at it, the new version of qemu7.2.0 is very similar to the previous version, some of which are just code additions and deletions, or structural changes between versions.
The function of qemu is essentially the same in all versions. Based on this point of view, panda can be updated to the new qemu version.
The panda project is very good, but due to the direction of the project and the old qemu, its panda activity is low.
Use vscode to open panda and qemu7.2.0 and open the current original qemu version of panda. Through the vscode search function, such as searching for the 'panda_' keyword, it is very easy to see where these keywords are used and update them.
Or use the text comparison tool to compare the differences between the three, plus the code reading ability of qemu, updating qemu7.2.0 is not a problem.

Regarding the qemu application, it can be compiled into so or other types of dynamic link libraries. One step forward is to optimize the panda project structure.

Project optimization + new version of qemu + core panda features = more active commits

**Comment by XVilka** [comment]
> We are actively developing a new version of PANDA atop qemu 8 using TCG plugins. We’re working with upstream to merge a new version of the PPP interface to allow plugins to interact with one another which we see as the first major blocker for PANDA-like functionality. After we hopefully get that merged, we’ll need to upstream patches to allow plugins to access guest registers and memory. Then we can start porting and upstreaming plugins as well. We currently have syscalls, osi_linux, and stringsearch ported to the new interface. And a limited pypanda interface. We haven’t yet tackled the record/replay system or LLVM IR as required by the taint system.
> 
> If you’re interested in joining the discussion with the QEMU developers, our latest patch series is at the following link. So far the discussion has only been between us and the maintainer, so other perspectives might be appreciated.
> 
> https://www.mail-archive.com/qemu-devel@nongnu.org/msg926042.html
> 
> Unfortunately our funding for this work runs out in a few days so we’re not sure what the future of this will be, but the work we've done so far is available [here](https://github.com/panda-re/qemu/tree/ppp_beta_with_plugins), and if upstream merges the QPP stuff we’ll try getting some of our API changes and plugins upstreamed too

Seems the patches were largely ignored. I suspect it was because developers were focused on 8.0 release. Now, when the merge window is open again, probably it makes sense to resend them? Or ping at the very least.

**EDITED**, nevermind, I see there was some initial feedback that remained unaddressed. Then it makes sense to send the second version of patches: https://lore.kernel.org/qemu-devel/20221213213757.4123265-1-fasano@mit.edu/

**Comment by XVilka** [comment]
@AndrewFasano you might be interested in this patch series, which adds reading registers from TCG plugins in a clean way (partially reviewed and approved): https://lore.kernel.org/qemu-devel/20230912071206.30751-1-akihiko.odaki@daynix.com/T/#t

**Comment by AndrewFasano** [comment]
Thanks for pointing that out, I think we could use that API to provide some example plugins to upstream with our PR! Should help a bunch and maybe we'll finally be able to get our initial changes merged :)

**Comment by cctv130** [comment]
@AndrewFasano I am dazzled by a large number of branches and intricate project structure, and I have lost confidence in submitting code for this project. Can we optimize this project structure, separate qemu, and keep pure panda?

**Comment by AndrewFasano** [comment]
@cctv130 I'd love to refactor PANDA to move things like the plugins into a different repo from the customized emulator code. I think it be easier to maintain the emulation code if all the plugin code/commits were kept separately. You can see a proposal I wrote in #947 and chime in there if you have any more specific suggestions. Perhaps we'll re-open the issue if there's sufficient interest. But for now, none of the other contributors to the project got on board with the idea so we haven't pursued it yet and the issue went stale. I'm hoping with [our ongoing qemu port](https://github.com/panda-re/qemu/tree/ppp_beta_plus_pr2.1) we'll do a better job keeping plugins independent from the core emulator logic.

In terms of branches, the only one you'd need to look at or think about is the main branch, `dev`. With the exception of changes we've made to the emulator logic, almost all PANDA code in the repo lives in the `panda` subdirectory.

**Comment by cctv130** [comment]
@AndrewFasano My suggestion is to start chat rooms such as telegram so that we can discuss problems together. I randomly looked at the submission records of the top contributors. in the last year, most contributors were very active in the initial period of panda, but it seems that panda has been abandoned in the past year. The reason for giving up is clear, if we follow the tcg route and tcg has no obstacles, I think we can immediately remove the messy branches and update the project structure. Thus giving a clear route to the new panda, attracting more developers to develop panda will not be a problem.

**Comment by cctv130** [comment]
@AndrewFasano 
I compared the panda and qemu 8.1 versions by deleting the panda folder under the panda project directory, and then directly searching for the original panda-related files. Although the two versions are very different, they have many similarities, which means that no matter how new qemu version is, The bottom layer of qemu basically changes little, even if there are changes, the idea of implementation is still the same, as others say that the new version has changed a lot, I also admit, but in the face of panda bottom layer, only a few qemu core calls, do not need to consider other qemu changes.
![1698240773526](https://github.com/panda-re/panda/assets/133784576/36b47ac4-86f4-4fa2-bd0d-aac96f7c48b2)

**Comment by AndrewFasano** [comment]
Superseded by #1383
